llvm/lib/Passes/PassBuilder.cpp

```diff
@@ -1877,6 +1894,33 @@ Error PassBuilder::parseLoopPass(LoopPassManager &LPM,
                                  inconvertibleErrorCode());
 }
 
+Error PassBuilder::parseMachinePass(MachineFunctionPassManager &MFPM,
+                                    const PipelineElement &E) {
+  StringRef Name = E.Name;
+  if (!E.InnerPipeline.empty())
+    return make_error<StringError>("invalid pipeline",
+                                   inconvertibleErrorCode());
+
+#define MACHINE_MODULE_PASS(NAME, PASS_NAME, CONSTRUCTOR)                      \
+  if (Name == NAME) {                                                          \
+    MFPM.addPass(PASS_NAME());                                                 \
+    return Error::success();                                                   \
+  }
```

Comment:
What does a machine module pass mean? If it's to mean module pass but for codegen, isn't `MFPM.addPass()` wrong as it assumes the MACHINE_MODULE_PASS pass to be a `MachineFunction` pass?

@paperchalice

Reply:
I think it's just a placeholder as there are no "MachineModulePasses"

Reply:
Oops the code is wrong here now. Initially when I took over this part, the original designer want to use machine function pass manager to manage both machine function pass and "machine module pass", but now I realized "machine module pass" is just a special kind of module pass that modifies all `MachineFunctionAnalysis::Result` i.e. `MachineFunciton`...

Reply:
I didn't get what modifies all `MachineFunctionAnalysis::Result` mean? I see that MachineModuleAnalysis and PassInstrumentationAnalysis are in this category.

Reply:
> I didn't get what modifies all `MachineFunctionAnalysis::Result` mean? I see that MachineModuleAnalysis and PassInstrumentationAnalysis are in this category.

Code is here: [`MachineFunctionAnalysis`](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/CodeGen/MachineFunctionAnalysis.h). It is just a simple wrapper of `MachineFunction` (may convert `MachineFunction` to a real analysis result in future).

Machine module pass can manipulate all machine functions via `FunctionAnalysisManager`. For example machine outliner will create some dummy IR functions and corresponding machine functions, then modify machine functions to compress the module size.

Reply:
I see, yes. To summarize, Machine module passes are Module passes that modify `MachineFunction`s via `MAM.getResult<Function...Proxy>(M)..getResult<MachineFunctionAnalysis>(F)`.